[core] Make preloading Jemalloc configurable for worker

Myasuka opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

The PR https://github.com/ray-project/ray/pull/39446 disables preloading Jemalloc for workers totally. However, Jemalloc is still useful in some cases, and we could make it configurable if user setting env RAY_LD_PRELOAD as 0.

The batch inference example code, using a TF model to infer the batch input of numpy's ndarray.
~~~python
ds = ray.data.read_tfrecords(xxx)
ds.map_batches(BatchPredictor)
.map_batches(BatchPostProcessor)
.write_parquet(path=output_path
~~~

I did a inference test with limited memory, and we can see the OOM counts decrease from 900+ to 700.
![image](https://github.com/user-attachments/assets/12e3f428-6b97-467d-801d-7ab4b9e3b659)




## Related issue number
Closes #47242

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

Myasuka commented:
@fishbone @rkooo567 could you please take a review?

rkooo567 reviewed (commented):
can you add a new section in this page to explain how to enable jemalloc for workers? https://docs.ray.io/en/master/ray-core/miscellaneous.html#tuning-ray-settings

hongchaodeng commented:
Please add your inference example and benchmark as well.

Myasuka commented:
> can you add a new section in this page to explain how to enable jemalloc for workers? https://docs.ray.io/en/master/ray-core/miscellaneous.html#tuning-ray-settings

Sure, I'll add a doc for this feature. However, why not adding the description in the [memory-profiling](https://docs.ray.io/en/master/ray-contribute/profiling.html#memory-profiling) doc?

Myasuka commented:
> Please add your inference example and benchmark as well.

I have added the example in the description, and I think the previous pasted image could describe the benchmark well.

rkooo567 commented:
> @rkooo567 Sure, I'll add a doc for this feature. However, why not adding the description in the [memory-profiling](https://docs.ray.io/en/master/ray-contribute/profiling.html#memory-profiling) doc?

I assume it is more of performance feature (that optimizes the memory usage). is enabling jemalloc allowing you to do mem profiling as well?

Myasuka commented:
> > @rkooo567 Sure, I'll add a doc for this feature. However, why not adding the description in the [memory-profiling](https://docs.ray.io/en/master/ray-contribute/profiling.html#memory-profiling) doc?
> 
> I assume it is more of performance feature (that optimizes the memory usage). is enabling jemalloc allowing you to do mem profiling as well?

I see, enabling Jemalloc for workers could also benefit for memory profiling, I will mentation this in the tuning and profiling page.

MissiontoMars commented:
We also need to enable jemalloc for the worker processes because some C++ code is wrapped into Python workers, jemalloc is very effective for optimizing the memory of this part of c++ code.

dayshah commented:
@Myasuka this makes sense to me, is this still something you want to merge?

Myasuka commented:
> @Myasuka this makes sense to me, is this still something you want to merge?

Sure, could you help to review and merge this PR? I will update this PR this week.

Myasuka committed:
[core] Make preloading Jemalloc configurable for worker

We can set `RAY_LD_PRELOAD_ON_WORKERS` as `true` with `RAY_JEMALLOC_LIB_PATH` and `RAY_JEMALLOC_PROFILE` provided to also preload jemalloc for worker.
This is a fix after #39446

Signed-off-by: Yun Tang <myasuka@live.com>

Myasuka commented:
@rkooo567 since https://docs.ray.io/en/master/ray-core/miscellaneous.html#tuning-ray-settings have been removed, I still add the docs in https://docs.ray.io/en/master/ray-contribute/profiling.html#memory-profiling

@dayshah Please also take a review on this update.

dstrodtman commented on doc/source/ray-contribute/profiling.rst:
```diff
@@ -88,11 +88,12 @@ Ray supports environment variables that override `LD_PRELOAD` on core components
 * `RAY_JEMALLOC_LIB_PATH`: The path to the jemalloc shared library `libjemalloc.so`
 * `RAY_JEMALLOC_CONF`: The MALLOC_CONF configuration for jemalloc, using comma-separated values. Read `jemalloc docs <http://jemalloc.net/jemalloc.3.html>`_ for more details.
 * `RAY_JEMALLOC_PROFILE`: Comma separated Ray components to run Jemalloc `.so`. e.g., ("raylet,gcs_server"). Note that the components should match the process type in `ray_constants.py`. (It means "RAYLET,GCS_SERVER" won't work).
+* `RAY_LD_PRELOAD`: If setting to `0`, Ray will also preload Jemalloc for worker when `RAY_JEMALLOC_LIB_PATH` and `RAY_JEMALLOC_PROFILE` provided. The default value is `1`, which means Ray won't preload Jemalloc for worker due to some library is not compatible with Jemalloc.
```
```suggestion
* `RAY_LD_PRELOAD`: Default value is `1`, which means Ray doesn't preload Jemalloc for workers if a library is incompatible with Jemalloc. Set to `0` to instruct Ray to preload Jemalloc for a worker using values configured by `RAY_JEMALLOC_LIB_PATH` and `RAY_JEMALLOC_PROFILE`.
```

Rewritten to improve clarity. Please ensure this is still correct.

"due to some library is not compatible with Jemalloc" unclear what this means. Does it never preload Jemalloc, or only fail to preload if it detects incompatibility?

dstrodtman reviewed (commented):
RAY_LD_PRELOAD description was not parsable. Can you review/iterate on my edit to make sure it's technically correct?

Docs should never contain future tense ("will") or passive voice (be verb with a past tense verb). Thanks!

dayshah reviewed (commented):
generally lgtm, can you address the doc comments

Also, I would prefer if we rename the `RAY_LD_PRELOAD` env variable to something like `RAY_LD_PRELOAD_ON_WORKERS`, which gets set to 0 by default. This makes it easier for a user to understand without looking through code and docs.

Myasuka commented:
I have updated this PR, please take a look @dayshah @dstrodtman

dayshah commented:
https://buildkite.com/ray-project/premerge/builds/41233
some lint failures

dayshah reviewed (approved):
ty! lgtm @edoakes @jjyao to merge

edoakes commented on python/ray/tests/test_advanced_4.py:
```diff
@@ -67,7 +69,7 @@ def test_jemalloc_env_var_propagate():
     """
     library_path = "/abc"
-    expected = {"LD_PRELOAD": library_path, "RAY_LD_PRELOAD": "1"}
+    expected = {"LD_PRELOAD": library_path, "RAY_LD_PRELOAD_ON_WORKERS": "false"}
```
let's stick to `"0"` and `"1"` for the values to match other env vars please

Myasuka commented on python/ray/_private/services.py:
```diff
@@ -227,7 +227,12 @@ def propagate_jemalloc_env_var(
     if not jemalloc_path:
         return {}
 
-    env_vars = {"LD_PRELOAD": jemalloc_path, "RAY_LD_PRELOAD": "1"}
```
I agree that `RAY_LD_PRELOAD_ON_WORKERS` looks better, if we still keep the undocumented `RAY_LD_PRELOAD`, it will make the code looks a bit weird.
> generally lgtm, can you address the doc comments
> 
> Also, I would prefer if we rename the `RAY_LD_PRELOAD` env variable to something like `RAY_LD_PRELOAD_ON_WORKERS`, which gets set to 0 by default. This makes it easier for a user to understand without looking through code and docs.

edoakes replied:
Did `RAY_LD_PRELOAD` _only_ apply to workers previously? I think that might be my gap in understanding. If that's the case, then this change LGTM. I was assuming that `RAY_LD_PRELOAD` also applied to system-level processes.

Myasuka replied:
YES, `RAY_LD_PRELOAD` was first introduced in https://github.com/ray-project/ray/pull/39446/files and only be used within that PR.
